<a href="https://colab.research.google.com/github/sanika-mhadgut/Deep_Learning/blob/master/Keras_Activation_(ii)_J031.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Name : Sanika Mhadgut
#Branch : Btech Data Science sem 6
#Roll No: J031


In [1]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='elu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='elu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Using TensorFlow backend.


11493376/11490434 [==============================] - 1s 0us/step





Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Train on 60000 samples, validate on 10000 samples
Epoch 1/10





60000/60000 [==============================] - 11s 179us/step - loss: 0.2887 - acc: 0.9115 - val_loss: 0.1566 - val_acc: 0.9511
Epoch 2/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.1222 - acc: 0.9628 - val_loss: 0.0947 - val_acc: 0.9692
Epoch 3/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0824 - acc: 0.9738 - val_loss: 0.0834 - val_acc: 0.9753
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0633 - acc: 0.9802 - val_loss: 0.0957 - val_acc: 0.9739
Epoch 5/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0491 - acc: 0.9842 - val_loss: 0.0942 - val_acc: 0.9733
Epoch 6/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0411 - acc

In [2]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softmax', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softmax'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 30us/step - loss: 2.2699 - acc: 0.2231 - val_loss: 2.1980 - val_acc: 0.5385
Epoch 2/10
60000/60000 [==============================] - 2s 25us/step - loss: 2.0445 - acc: 0.7262 - val_loss: 1.8543 - val_acc: 0.7560
Epoch 3/10
60000/60000 [==============================] - 2s 26us/step - loss: 1.6408 - acc: 0.7599 - val_loss: 1.4194 - val_acc: 0.7602
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 1.2237 - acc: 0.7642 - val_loss: 1.0370 - val_acc: 0.7668
Epoch 5/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.9010 - acc: 0.7695 - val_loss: 0.7820 - val_acc: 0.7702
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.7041 - acc: 0.7761 - val_loss: 0.6386 - val_acc: 0.7761
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.5996 - acc: 0.7816 - val_loss: 0.5684 - val_acc

In [3]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='selu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='selu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.3293 - acc: 0.9012 - val_loss: 0.1460 - val_acc: 0.9553
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.1363 - acc: 0.9583 - val_loss: 0.1322 - val_acc: 0.9593
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0954 - acc: 0.9702 - val_loss: 0.0926 - val_acc: 0.9737
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0731 - acc: 0.9772 - val_loss: 0.1185 - val_acc: 0.9652
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0581 - acc: 0.9816 - val_loss: 0.1134 - val_acc: 0.9666
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0514 - acc: 0.9833 - val_loss: 0.0941 - val_acc: 0.9719
Epoch 7/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0423 - acc: 0.9865 - val_loss: 0.0878 - val_acc

In [4]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softplus', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softplus'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.4522 - acc: 0.8608 - val_loss: 0.2727 - val_acc: 0.9107
Epoch 2/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.1711 - acc: 0.9478 - val_loss: 0.1648 - val_acc: 0.9454
Epoch 3/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.1114 - acc: 0.9656 - val_loss: 0.0988 - val_acc: 0.9691
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0836 - acc: 0.9743 - val_loss: 0.0870 - val_acc: 0.9740
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0652 - acc: 0.9799 - val_loss: 0.0788 - val_acc: 0.9778
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0532 - acc: 0.9832 - val_loss: 0.0780 - val_acc: 0.9786
Epoch 7/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0437 - acc: 0.9867 - val_loss: 0.0844 - val_acc

In [5]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softsign', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softsign'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.2982 - acc: 0.9116 - val_loss: 0.2115 - val_acc: 0.9340
Epoch 2/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.1444 - acc: 0.9565 - val_loss: 0.1254 - val_acc: 0.9618
Epoch 3/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0996 - acc: 0.9691 - val_loss: 0.0930 - val_acc: 0.9700
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0741 - acc: 0.9772 - val_loss: 0.0891 - val_acc: 0.9716
Epoch 5/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.0578 - acc: 0.9825 - val_loss: 0.0810 - val_acc: 0.9747
Epoch 6/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0455 - acc: 0.9859 - val_loss: 0.0875 - val_acc: 0.9720
Epoch 7/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.0352 - acc: 0.9893 - val_loss: 0.0690 - val_acc

In [6]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='relu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.2282 - acc: 0.9307 - val_loss: 0.1051 - val_acc: 0.9671
Epoch 2/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0838 - acc: 0.9743 - val_loss: 0.0728 - val_acc: 0.9778
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0545 - acc: 0.9836 - val_loss: 0.0661 - val_acc: 0.9799
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0386 - acc: 0.9877 - val_loss: 0.0727 - val_acc: 0.9800
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0297 - acc: 0.9906 - val_loss: 0.0915 - val_acc: 0.9768
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0226 - acc: 0.9930 - val_loss: 0.0877 - val_acc: 0.9786
Epoch 7/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0188 - acc: 0.9942 - val_loss: 0.0746 - val_acc

In [7]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='tanh', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='tanh'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.2856 - acc: 0.9137 - val_loss: 0.1541 - val_acc: 0.9540
Epoch 2/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.1217 - acc: 0.9628 - val_loss: 0.0981 - val_acc: 0.9699
Epoch 3/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0793 - acc: 0.9755 - val_loss: 0.0791 - val_acc: 0.9757
Epoch 4/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0563 - acc: 0.9829 - val_loss: 0.0769 - val_acc: 0.9755
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0413 - acc: 0.9877 - val_loss: 0.0670 - val_acc: 0.9800
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0302 - acc: 0.9916 - val_loss: 0.0850 - val_acc: 0.9744
Epoch 7/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0229 - acc: 0.9936 - val_loss: 0.0753 - val_acc

In [8]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.4957 - acc: 0.8563 - val_loss: 0.2480 - val_acc: 0.9239
Epoch 2/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.2178 - acc: 0.9347 - val_loss: 0.1888 - val_acc: 0.9422
Epoch 3/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.1554 - acc: 0.9526 - val_loss: 0.1526 - val_acc: 0.9532
Epoch 4/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.1166 - acc: 0.9650 - val_loss: 0.1311 - val_acc: 0.9596
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0921 - acc: 0.9715 - val_loss: 0.0973 - val_acc: 0.9691
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0748 - acc: 0.9769 - val_loss: 0.0903 - val_acc: 0.9727
Epoch 7/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0617 - acc: 0.9806 - val_loss: 0.0729 - val_acc

In [9]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='hard_sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='hard_sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.5053 - acc: 0.8537 - val_loss: 0.2608 - val_acc: 0.9228
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.2155 - acc: 0.9356 - val_loss: 0.1885 - val_acc: 0.9410
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.1496 - acc: 0.9547 - val_loss: 0.1399 - val_acc: 0.9573
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1110 - acc: 0.9665 - val_loss: 0.1162 - val_acc: 0.9627
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0865 - acc: 0.9740 - val_loss: 0.1011 - val_acc: 0.9696
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0689 - acc: 0.9791 - val_loss: 0.0975 - val_acc: 0.9697
Epoch 7/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0563 - acc: 0.9828 - val_loss: 0.0805 - val_acc

In [10]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='exponential', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='exponential'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 35us/step - loss: 14.5313 - acc: 0.0975 - val_loss: 14.5482 - val_acc: 0.0974
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 14.5463 - acc: 0.0975 - val_loss: 14.5482 - val_acc: 0.0974
Epoch 3/10
60000/60000 [==============================] - 2s 26us/step - loss: 14.5463 - acc: 0.0975 - val_loss: 14.5482 - val_acc: 0.0974
Epoch 4/10
60000/60000 [==============================] - 2s 27us/step - loss: 14.5463 - acc: 0.0975 - val_loss: 14.5482 - val_acc: 0.0974
Epoch 5/10
60000/60000 [==============================] - 2s 26us/step - loss: 14.5463 - acc: 0.0975 - val_loss: 14.5482 - val_acc: 0.0974
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 14.5463 - acc: 0.0975 - val_loss: 14.5482 - val_acc: 0.0974
Epoch 7/10
60000/60000 [==============================] - 2s 26us/step - loss: 14.5463 - acc: 0.0975 - val_loss: 14.

In [11]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='linear', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='linear'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 37us/step - loss: 0.4269 - acc: 0.8761 - val_loss: 0.3732 - val_acc: 0.8920
Epoch 2/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.3422 - acc: 0.9032 - val_loss: 0.3192 - val_acc: 0.9079
Epoch 3/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.3207 - acc: 0.9088 - val_loss: 0.3255 - val_acc: 0.9137
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.3090 - acc: 0.9148 - val_loss: 0.3155 - val_acc: 0.9070
Epoch 5/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.3004 - acc: 0.9157 - val_loss: 0.2984 - val_acc: 0.9188
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.2932 - acc: 0.9178 - val_loss: 0.3497 - val_acc: 0.9001
Epoch 7/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.2898 - acc: 0.9194 - val_loss: 0.3367 - val_acc

In [12]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.LeakyReLU(alpha=0.3), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.LeakyReLU(alpha=0.3)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as LeakyReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.2608 - acc: 0.9208 - val_loss: 0.1469 - val_acc: 0.9546
Epoch 2/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.1257 - acc: 0.9613 - val_loss: 0.1058 - val_acc: 0.9667
Epoch 3/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0910 - acc: 0.9716 - val_loss: 0.1079 - val_acc: 0.9676
Epoch 4/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0756 - acc: 0.9764 - val_loss: 0.0857 - val_acc: 0.9732
Epoch 5/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0641 - acc: 0.9804 - val_loss: 0.1289 - val_acc: 0.9626
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0551 - acc: 0.9826 - val_loss: 0.1123 - val_acc: 0.9696
Epoch 7/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0482 - acc: 0.9846 - val_loss: 0.1049 - val_acc

In [13]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as PReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.2309 - acc: 0.9285 - val_loss: 0.1026 - val_acc: 0.9670
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0838 - acc: 0.9742 - val_loss: 0.0881 - val_acc: 0.9725
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0550 - acc: 0.9828 - val_loss: 0.0715 - val_acc: 0.9797
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0397 - acc: 0.9882 - val_loss: 0.0986 - val_acc: 0.9728
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0293 - acc: 0.9912 - val_loss: 0.0795 - val_acc: 0.9791
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0225 - acc: 0.9929 - val_loss: 0.0793 - val_acc: 0.9807
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0186 - acc: 0.9942 - val_loss: 0.0853 - val_acc

In [14]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ELU(alpha=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ELU(alpha=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ELU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 41us/step - loss: 0.2873 - acc: 0.9131 - val_loss: 0.1646 - val_acc: 0.9484
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.1192 - acc: 0.9638 - val_loss: 0.1371 - val_acc: 0.9573
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0810 - acc: 0.9755 - val_loss: 0.0997 - val_acc: 0.9689
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0620 - acc: 0.9807 - val_loss: 0.0783 - val_acc: 0.9747
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0494 - acc: 0.9836 - val_loss: 0.0862 - val_acc: 0.9759
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0400 - acc: 0.9874 - val_loss: 0.0748 - val_acc: 0.9788
Epoch 7/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0327 - acc: 0.9897 - val_loss: 0.0769 - val_acc

In [15]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ThresholdedReLU(theta=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ThresholdedReLU(theta=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ThresholdedReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 0.5790 - acc: 0.8121 - val_loss: 0.1989 - val_acc: 0.9393
Epoch 2/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.1557 - acc: 0.9527 - val_loss: 0.1437 - val_acc: 0.9558
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.1026 - acc: 0.9689 - val_loss: 0.1018 - val_acc: 0.9694
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0773 - acc: 0.9765 - val_loss: 0.0928 - val_acc: 0.9738
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0612 - acc: 0.9811 - val_loss: 0.0869 - val_acc: 0.9752
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0502 - acc: 0.9842 - val_loss: 0.0868 - val_acc: 0.9759
Epoch 7/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0417 - acc: 0.9869 - val_loss: 0.0858 - val_acc

In [16]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.Softmax(axis=-1), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.Softmax(axis=-1)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as Softmax) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 2.2680 - acc: 0.1941 - val_loss: 2.1927 - val_acc: 0.5537
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 2.0365 - acc: 0.7149 - val_loss: 1.8461 - val_acc: 0.7610
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 1.6343 - acc: 0.7576 - val_loss: 1.4130 - val_acc: 0.7606
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 1.2197 - acc: 0.7623 - val_loss: 1.0329 - val_acc: 0.7681
Epoch 5/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.8988 - acc: 0.7691 - val_loss: 0.7774 - val_acc: 0.7677
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.7016 - acc: 0.7737 - val_loss: 0.6372 - val_acc: 0.7781
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.5986 - acc: 0.7802 - val_loss: 0.5689 - val_acc

In [17]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 0.2246 - acc: 0.9318 - val_loss: 0.1237 - val_acc: 0.9628
Epoch 2/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0826 - acc: 0.9744 - val_loss: 0.0747 - val_acc: 0.9770
Epoch 3/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0542 - acc: 0.9834 - val_loss: 0.0720 - val_acc: 0.9801
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0387 - acc: 0.9877 - val_loss: 0.0904 - val_acc: 0.9737
Epoch 5/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0292 - acc: 0.9910 - val_loss: 0.0825 - val_acc: 0.9786
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0221 - acc: 0.9932 - val_loss: 0.0812 - val_acc: 0.9808
Epoch 7/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0175 - acc: 0.9943 - val_loss: 0.0774 - val_acc

relu

Test loss: 0.08459654634964409

Test accuracy: 0.9799

softmax

Test loss: 0.39845509810447693

Test accuracy: 0.8642

selu

Test loss: 0.11848029121595427

Test accuracy: 0.9726

softplus

Test loss: 0.08013557475531798

Test accuracy: 0.9798

softsign

Test loss: 0.07156896027913608

Test accuracy: 0.9793

relu

Test loss: 0.09273777272929686

Test accuracy: 0.9843

tanh

Test loss: 0.06885148700817954

Test accuracy: 0.9803

sigmoid

Test loss: 0.07136517209208104

Test accuracy: 0.9781

hard_sigmoid

Test loss: 0.06968761095895898

Test accuracy: 0.9795

exponential

Test loss: 14.573981648254394

Test accuracy: 0.0958

linear

Test loss: 0.30165807257443666

Test accuracy: 0.9158

LeakyReLU

Test loss: 0.11553157684192265

Test accuracy: 0.9709

PReLU

Test loss: 0.11553157684192265

Test accuracy: 0.9709

ELU

Test loss: 0.1115254690002792

Test accuracy: 0.9743

ThresholdedReLU

Test loss: 0.08483904730657359

Test accuracy: 0.9786

Softmax

Test loss: 0.2818538703083992

Test accuracy: 0.9333

ReLU

Test loss: 0.09205757192509677

Test accuracy: 0.9829

In [18]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import fashion_mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.layers import Activation, Flatten, Conv2D

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], 28, 28, 1))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], 28, 28, 1))
X_test = X_test.astype('float32') / 255



# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)

# Build neural network
model = models.Sequential()
model.add(Conv2D(512, (28, 28), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

4423680/4422102 [==============================] - 1s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 7s 122us/step - loss: 0.5945 - acc: 0.7764 - val_loss: 0.4231 - val_acc: 0.8499
Epoch 2/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.3963 - acc: 0.8526 - val_loss: 0.4887 - val_acc: 0.8187
Epoch 3/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.3524 - acc: 0.8698 - val_loss: 0.3748 - val_acc: 0.8619
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.3274 - acc: 0.8804 - val_loss: 0.3919 - val_acc: 0.8654
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.3093 - acc: 0.8860 - val_loss: 0.3836 - val_acc: 0.8664
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.2968 - acc: 0.8897 - val_loss: 0.3675 - val_acc: 0.8741
Epoch 7/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.2897 - acc: 0.8938 - val_loss: 0.4319 - val_ac

FASHION_MNIST

Test loss: 0.4138208844304085

Test accuracy: 0.8699

In [19]:
import keras
from keras import models
from keras.layers import Dense, Dropout, Conv2D,Flatten
from keras.utils import to_categorical
from keras.datasets import cifar10
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000,32,32,3)
X_test = X_test.reshape(10000,32,32,3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)


# Build neural network
model = models.Sequential()
model.add(Conv2D(1024, (32, 32), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(32 * 32,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

170500096/170498071 [==============================] - 4s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 4s 84us/step - loss: 2.0291 - acc: 0.2946 - val_loss: 1.7651 - val_acc: 0.3785
Epoch 2/10
50000/50000 [==============================] - 3s 63us/step - loss: 1.7035 - acc: 0.3861 - val_loss: 1.6622 - val_acc: 0.3911
Epoch 3/10
50000/50000 [==============================] - 3s 60us/step - loss: 1.6072 - acc: 0.4265 - val_loss: 1.5729 - val_acc: 0.4380
Epoch 4/10
50000/50000 [==============================] - 3s 63us/step - loss: 1.5379 - acc: 0.4497 - val_loss: 1.5461 - val_acc: 0.4469
Epoch 5/10
50000/50000 [==============================] - 3s 60us/step - loss: 1.4885 - acc: 0.4676 - val_loss: 1.4870 - val_acc: 0.4749
Epoch 6/10
50000/50000 [==============================] - 3s 61us/step - loss: 1.4494 - acc: 0.4859 - val_loss: 1.4456 - val_acc: 0.4874
Epoch 7/10
50000/50000 [==============================] - 3s 61us/step - loss: 1.4089 - acc: 0.4985 - val_loss: 1.4785 - val_acc

CIFAR10

Test loss: 1.4009930366516112

Test accuracy: 0.4971

In [20]:
import keras
from keras import models
from keras.layers import Dense, Dropout, Conv2D,Flatten
from keras.utils import to_categorical
from keras.datasets import cifar100
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers

#import dataset
(X_train, y_train), (X_test, y_test) = cifar100.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000,32,32,3)
X_test = X_test.reshape(10000,32,32,3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 100)
Y_test =  to_categorical(y_test, 100)

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)


# Build neural network
model = models.Sequential()
model.add(Conv2D(1024, (32, 32), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(32 * 32,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(100, activation='softmax'))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

169009152/169001437 [==============================] - 4s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 4s 83us/step - loss: 4.1830 - acc: 0.0649 - val_loss: 3.9228 - val_acc: 0.0939
Epoch 2/10
50000/50000 [==============================] - 3s 61us/step - loss: 3.7574 - acc: 0.1248 - val_loss: 3.6586 - val_acc: 0.1526
Epoch 3/10
50000/50000 [==============================] - 3s 62us/step - loss: 3.5747 - acc: 0.1564 - val_loss: 3.5700 - val_acc: 0.1609
Epoch 4/10
50000/50000 [==============================] - 3s 61us/step - loss: 3.4468 - acc: 0.1804 - val_loss: 3.4474 - val_acc: 0.1873
Epoch 5/10
50000/50000 [==============================] - 3s 61us/step - loss: 3.3369 - acc: 0.2005 - val_loss: 3.4000 - val_acc: 0.1973
Epoch 6/10
50000/50000 [==============================] - 3s 64us/step - loss: 3.2321 - acc: 0.2203 - val_loss: 3.3279 - val_acc: 0.2121
Epoch 7/10
50000/50000 [==============================] - 3s 63us/step - loss: 3.1498 - acc: 0.2352 - val_loss: 3.3126 - val_acc

CIFAR100

Test loss: 3.242277897644043

Test accuracy: 0.2326